File: ios-native/Leyne/AppOpenAd.swift
// Google AdMob App Open ad — full-screen ad shown when the app returns to the
// foreground from the background. The single highest-value format for a
// high-frequency utility app like a bus tracker (opened many times a day).
//
// UX-first / policy-safe guards, all enforced in `showIfReady(model:)`:
//   • Respects the same gates as the banner: `AdConfig.adsEnabled`,
//     `AdConfig.screenshotMode`, and `AdConfig.started` (UMP + ATT + SDK init).
//   • NEVER during/before onboarding, or over the launch splash.
//   • NEVER when a notification / widget / Spotlight tap brought the user to
//     the foreground — they want their bus, not an ad (`suppressNextPresentation`).
//   • At most once every `minInterval` (frequency cap, persisted in UserDefaults).
//   • Only a fresh creative — App Open ads expire `maxCacheAge` after load
//     (AdMob policy); stale ones are dropped and reloaded.
//
// Wired to fire on WARM foreground only (scenePhase .background → .active in
// LeyneApp), never on cold launch — least jarring, and avoids racing the
// consent / ATT prompts that run on first launch.

import SwiftUI
import GoogleMobileAds
import os

private let aoaLog = Logger(subsystem: "com.leyne.Leyne", category: "AppOpenAd")

enum AppOpenAdConfig {
    // Ad unit selection mirrors AdConfig.bannerUnitID:
    //   • DEBUG (Xcode Run): Google's reserved App Open TEST unit.
    //   • RELEASE: the real production unit, unless `forceTestUnitForRelease`
    //     (the TestFlight toggle in AdConfig) forces the test unit.
    //
    // Production App Open unit lives in the Leyne AdMob account
    // (leyne0000@gmail.com, ca-app-pub-5864511655536507), matching the
    // GADApplicationIdentifier (…~6330743279) in LeyneInfo.plist.
    #if DEBUG
    static let unitID = "ca-app-pub-3940256099942544/5575463023"  // Google test
    #else
    static let unitID = AdConfig.forceTestUnitForRelease
        ? "ca-app-pub-3940256099942544/5575463023"   // Google test
        : "ca-app-pub-5864511655536507/4093216883"   // leyne-acct prod App Open
    #endif

    /// Frequency cap — at most one App Open ad per this window.
    static let minInterval: TimeInterval = 5 * 60

    /// App Open creatives expire 4h after load (AdMob). Drop + reload past this.
    static let maxCacheAge: TimeInterval = 4 * 60 * 60
}

@MainActor
final class AppOpenAdManager: NSObject {
    static let shared = AppOpenAdManager()

    private var ad: AppOpenAd?
    private var loadTime: Date = .distantPast
    private var isLoading = false
    private var isShowing = false
    private var suppressUntil: Date = .distantPast
    private var wasBackgrounded = false
    private let lastShownKey = "leyne.appOpenAd.lastShown"

    private override init() { super.init() }

    /// Call when the scene enters the background. scenePhase steps through
    /// .inactive between .background and .active, so the previous phase is NOT
    /// reliably .background on return — we track backgrounding explicitly here.
    /// Also preloads so an ad is ready for the return-to-foreground.
    func noteBackgrounded() {
        wasBackgrounded = true
        preload()
    }

    /// Call when the scene becomes active. Presents only if we actually returned
    /// from the background — ignores cold launch (never hit .background) and
    /// transient .inactive blips (Control Center, the notification shade).
    func showIfReturningToForeground(model: AppModel) {
        guard wasBackgrounded else { return }
        wasBackgrounded = false
        showIfReady(model: model)
    }

    /// Called by the notification / widget / Spotlight handlers so the App Open
    /// ad is skipped on the foreground they triggered. Short TTL covers the race
    /// between the scenePhase transition and the deep-link handler firing.
    func suppressNextPresentation() {
        suppressUntil = Date().addingTimeInterval(3)
        aoaLog.notice("App Open suppressed (deep link)")
    }

    private var isExpired: Bool {
        Date().timeIntervalSince(loadTime) > AppOpenAdConfig.maxCacheAge
    }

    /// Load a single ad ahead of time if we don't already have a fresh one.
    func preload() {
        guard AdConfig.adsEnabled, !AdConfig.screenshotMode else { return }
        guard AdConfig.started else { return }       // consent + SDK ready
        guard !isLoading, ad == nil || isExpired else { return }
        isLoading = true
        aoaLog.notice("App Open loading…")
        AppOpenAd.load(with: AppOpenAdConfig.unitID, request: Request()) { [weak self] ad, error in
            // GADAppOpenAd delivers this completion on the main queue; the
            // manager is @MainActor so touching its state here is safe (same
            // pattern as the banner's delegate callbacks).
            guard let self else { return }
            self.isLoading = false
            if let error {
                aoaLog.error("App Open load failed: \(error.localizedDescription)")
                return
            }
            ad?.fullScreenContentDelegate = self
            self.ad = ad
            self.loadTime = Date()
            aoaLog.notice("App Open loaded")
        }
    }

    /// Present the App Open ad on foreground IF every guard passes; otherwise
    /// make sure one is loading for next time. Safe to call on every foreground.
    func showIfReady(model: AppModel) {
        guard AdConfig.adsEnabled, !AdConfig.screenshotMode else { return }
        guard AdConfig.started else { preload(); return }

        // Never during/before onboarding, or over the launch splash.
        guard model.onboarded, !model.showOnboarding, !model.launching else {
            aoaLog.notice("App Open skip: onboarding/launch")
            return
        }
        // A notification / widget / Spotlight tap brought us here — show content.
        guard Date() >= suppressUntil else {
            aoaLog.notice("App Open skip: deep-link suppression")
            return
        }
        // Frequency cap.
        let last = Date(timeIntervalSince1970:
            UserDefaults.standard.double(forKey: lastShownKey))
        guard Date().timeIntervalSince(last) >= AppOpenAdConfig.minInterval else {
            aoaLog.notice("App Open skip: within frequency cap")
            return
        }

        guard !isShowing else { return }
        // No fresh ad ready — drop any stale one and load for next time.
        guard let ad, !isExpired else {
            if isExpired { self.ad = nil }
            preload()
            return
        }
        guard let root = Self.rootVC() else { return }

        isShowing = true
        // Record BEFORE presenting so a present-failure still counts against the
        // cap (don't hammer the user with retries).
        UserDefaults.standard.set(Date().timeIntervalSince1970, forKey: lastShownKey)
        self.ad = nil   // hand ownership to the present() lifecycle
        aoaLog.notice("App Open present")
        ad.present(from: root)
    }

    static func rootVC() -> UIViewController? {
        UIApplication.shared.connectedScenes
            .compactMap { $0 as? UIWindowScene }
            .flatMap(\.windows)
            .first(where: \.isKeyWindow)?
            .rootViewController
    }
}

extension AppOpenAdManager: FullScreenContentDelegate {
    func adDidDismissFullScreenContent(_ ad: FullScreenPresentingAd) {
        isShowing = false
        self.ad = nil
        preload()   // get the next one ready
    }

    func ad(_ ad: FullScreenPresentingAd,
            didFailToPresentFullScreenContentWithError error: Error) {
        aoaLog.error("App Open present failed: \(error.localizedDescription)")
        isShowing = false
        self.ad = nil
        preload()
    }
}

File: lib/services/app_open_ad.dart
// Google AdMob App Open ad — full-screen ad shown when the app is launched or
// returns to the foreground. The single highest-value format for a
// high-frequency utility app like a bus tracker (opened many times a day).
// AdMob: "apps with frequent opens (>once every 4h) see the best performance."
//
// UX-first / policy-safe guards, all enforced in [showIfAvailable]:
//   • Respects the same master switches as the banner: kLyneAdsEnabled,
//     kLyneScreenshotMode, and AdConsent.started (UMP + SDK init done).
//   • NEVER during or before onboarding (AppModel.onboardingDone gate).
//   • NEVER when a notification / deep-link tap brought the user to the
//     foreground — they want their bus, not an ad ([suppressNext]).
//   • At most once every [_minInterval] (frequency cap, persisted).
//   • Only shows a fresh creative — App Open ads expire [_maxCacheAge] after
//     load (AdMob policy), so stale ones are dropped and reloaded.
//   • Never stacks with the Interstitial (shared FullScreenAdGate) — AdMob
//     disallows an ad immediately before/after an app-open ad.
//
// Fires on BOTH a cold launch ([showOnColdLaunch], for returning users — never
// the very first launch) and warm foreground (AppLifecycleListener onResume).
// The 24h cap means brief in-and-out app-switching won't trigger one; only a
// genuine new session (a return after a long break) can, at most once a day.

import 'dart:async';

import 'package:flutter/foundation.dart';
import 'package:google_mobile_ads/google_mobile_ads.dart';
import 'package:shared_preferences/shared_preferences.dart';

import '../state/app_model.dart';
import '../widgets/ad_banner.dart'
    show kLyneAdsEnabled, kLyneScreenshotMode, kLyneAdsTest;
import 'ad_consent.dart';
import 'full_screen_ad_gate.dart';

class AppOpenAdManager {
  AppOpenAdManager._();
  static final AppOpenAdManager instance = AppOpenAdManager._();

  // Ad unit selection mirrors the banner (ad_banner.dart `_bannerUnitId`):
  //   • DEBUG or LYNE_ADS_TEST=true → Google's reserved App Open TEST unit.
  //   • RELEASE (public Play Store) → the real production App Open unit (Leyne
  //     AdMob account, leyne0000@gmail.com, ca-app-pub-5864511655536507,
  //     matching the Android app id …~5685985257).
  static const String _testUnit = 'ca-app-pub-3940256099942544/9257395921';
  static const String _prodUnit =
      'ca-app-pub-5864511655536507/1467053541'; // leyne-acct prod App Open

  String get _unitId => (kDebugMode || kLyneAdsTest) ? _testUnit : _prodUnit;

  /// When false, NO App Open ad is shown on a cold launch — opening the app
  /// never greets the user with a full-screen ad. Set false on tester feedback
  /// that launch ads were too aggressive (QOL: "reduce advertisement esp upon
  /// app launch"). Warm foreground returns after the [_minInterval] window can
  /// still show one (rare), so some revenue is preserved. Flip to true to
  /// restore cold-launch presentation.
  static const bool _coldLaunchEnabled = false;

  /// Frequency cap — at most one App Open ad per this window. Raised 4h → 6h →
  /// 24h as tester feedback kept flagging the warm-return launch ad as annoying
  /// (a bus app is opened many times a day; an ad on re-entry interrupts exactly
  /// when the user wants arrivals). At 24h it shows at most once per day, on a
  /// genuine new session after a long break. Cold-launch stays disabled
  /// ([_coldLaunchEnabled]); banners + the interstitial-on-exit carry the rest.
  static const Duration _minInterval = Duration(hours: 24);

  /// App Open creatives expire 4h after load (AdMob). Drop + reload past this.
  static const Duration _maxCacheAge = Duration(hours: 4);

  static const String _lastShownKey = 'lyne.appOpenAd.lastShownMs';

  /// Set true after the first launch into the main UI, so the cold-launch ad
  /// never greets a brand-new user on their very first open (AdMob guidance).
  static const String _coldLaunchedKey = 'lyne.appOpenAd.coldLaunchedBefore';

  AppOpenAd? _ad;
  DateTime? _loadTime;
  bool _isLoading = false;
  bool _isShowing = false;
  DateTime _suppressUntil = DateTime.fromMillisecondsSinceEpoch(0);

  /// Called by the notification-tap and deep-link handlers so the App Open ad
  /// is skipped on the foreground they triggered (short TTL covers the race
  /// between the lifecycle resume and the deep-link callback firing).
  void suppressNext() {
    _suppressUntil = DateTime.now().add(const Duration(seconds: 3));
  }

  bool get _isExpired =>
      _loadTime == null || DateTime.now().difference(_loadTime!) > _maxCacheAge;

  /// Load a single ad ahead of time if we don't already have a fresh one.
  void preload() {
    if (!kLyneAdsEnabled || kLyneScreenshotMode) return;
    if (!AdConsent.started) return; // consent + MobileAds.initialize must be done
    if (_isLoading || (_ad != null && !_isExpired)) return;
    _isLoading = true;
    AppOpenAd.load(
      adUnitId: _unitId,
      request: const AdRequest(),
      adLoadCallback: AppOpenAdLoadCallback(
        onAdLoaded: (ad) {
          _ad = ad;
          _loadTime = DateTime.now();
          _isLoading = false;
        },
        onAdFailedToLoad: (error) {
          _isLoading = false;
          if (kDebugMode) {
            // ignore: avoid_print
            print('[ads] app-open load failed: ${error.message}');
          }
        },
      ),
    );
  }

  /// Preload once consent has resolved. Polls lightly (the banner uses the same
  /// pattern) since consent is fired fire-and-forget from main(). Bounded so it
  /// never loops forever when ads are disabled or consent stalls.
  void preloadWhenReady([int attemptsLeft = 15]) {
    if (!kLyneAdsEnabled || kLyneScreenshotMode) return;
    if (AdConsent.started) {
      preload();
      return;
    }
    if (attemptsLeft <= 0) return;
    Future.delayed(
      const Duration(milliseconds: 800),
      () => preloadWhenReady(attemptsLeft - 1),
    );
  }

  /// Cold-launch presentation: on a genuine app launch (not a warm resume),
  /// show one App Open ad once it's loaded — bounded to a short window after
  /// launch so a late creative never covers content the user is already using.
  /// Skipped on the very first launch into the app (AdMob guidance: don't greet
  /// a brand-new user with an ad). All the usual guards (4h cap, onboarding,
  /// deep-link suppression, cross-format gap) still apply inside
  /// [showIfAvailable].
  Future<void> showOnColdLaunch() async {
    if (!kLyneAdsEnabled || kLyneScreenshotMode) return;
    // Cold-launch App Open ads are disabled (see [_coldLaunchEnabled]) so the
    // app never opens straight into an ad. We still mark the first-launch flag
    // below so flipping the toggle back on later behaves correctly.
    if (!_coldLaunchEnabled) return;
    final prefs = await SharedPreferences.getInstance();
    if (!(prefs.getBool(_coldLaunchedKey) ?? false)) {
      await prefs.setBool(_coldLaunchedKey, true); // first launch — no ad
      return;
    }
    // Wait briefly for consent + a loaded creative, then present. Bounded
    // (~5s) so we never show over content the user has already started using.
    for (var i = 0; i < 10; i++) {
      if (AdConsent.started && _ad != null && !_isExpired) {
        await showIfAvailable();
        return;
      }
      await Future.delayed(const Duration(milliseconds: 500));
    }
  }

  /// Show the App Open ad on foreground IF every guard passes; otherwise make
  /// sure one is loading for next time. Safe to call on every resume.
  Future<void> showIfAvailable() async {
    if (!kLyneAdsEnabled || kLyneScreenshotMode) return;
    if (!AdConsent.started) {
      preload();
      return;
    }
    // Never during/before onboarding.
    if (!AppModel.shared.onboardingDone) return;
    // A notification / deep link brought us here — show content, not an ad.
    if (DateTime.now().isBefore(_suppressUntil)) return;

    // Frequency cap.
    final prefs = await SharedPreferences.getInstance();
    final lastMs = prefs.getInt(_lastShownKey) ?? 0;
    final last = DateTime.fromMillisecondsSinceEpoch(lastMs);
    if (DateTime.now().difference(last) < _minInterval) return;

    // Shared cross-format gap — don't stack right after an Interstitial.
    if (!await FullScreenAdGate.gapElapsed()) return;

    if (_isShowing) return;
    // No fresh ad ready — drop any stale one and load for next time.
    if (_ad == null || _isExpired) {
      _ad?.dispose();
      _ad = null;
      preload();
      return;
    }

    final ad = _ad!;
    ad.fullScreenContentCallback = FullScreenContentCallback(
      onAdShowedFullScreenContent: (_) => _isShowing = true,
      onAdDismissedFullScreenContent: (ad) {
        _isShowing = false;
        ad.dispose();
        _ad = null;
        preload(); // ready for next time
      },
      onAdFailedToShowFullScreenContent: (ad, error) {
        _isShowing = false;
        ad.dispose();
        _ad = null;
        preload();
        if (kDebugMode) {
          // ignore: avoid_print
          print('[ads] app-open show failed: ${error.message}');
        }
      },
    );
    // Record BEFORE showing so a present-failure still counts against the cap
    // (don't hammer the user with retries).
    await prefs.setInt(_lastShownKey, DateTime.now().millisecondsSinceEpoch);
    await FullScreenAdGate.markShown();
    _ad = null; // hand ownership to the show() lifecycle
    ad.show();
  }
}
